[mod] package.html template: additional links (a python dict)

dkuku opened the pull request:
## What does this PR do?
Refactor package template to include custom amount of links
Rename publishedDate variable to published_date
change engines that use this variable
<!-- MANDATORY -->
![image](https://github.com/searxng/searxng/assets/904179/a8e54e4f-8bb8-42e0-8651-d65711f76a08)
![image](https://github.com/searxng/searxng/assets/904179/544b647f-5c63-40e3-8d10-210fd3197d16)

<!-- explain the changes in your PR, algorithms, design, architecture -->

## Why is this change important?

<!-- MANDATORY -->

<!-- explain the motivation behind your PR -->

## How to test this PR locally?

<!-- commands to run the tests or instructions to test the changes -->

## Author's checklist

<!-- additional notes for reviewers -->

## Related issues

<!--
Closes #234
-->

dkuku committed:
[mod] package.html template: additional links (a python dict)

- Closes: https://github.com/searxng/searxng/issues/3456

Bnyro commented on searx/engines/hex.py:
```diff
@@ -46,10 +46,9 @@ def response(resp):
-                "tags": tags,
-                "homepage": meta.get("links", {}).get("homepage"),
-                "source_code_url": meta.get("links", {}).get("github"),
+                "published_date": published_date,
```
Please don't change that, it won't continue to work that way. The field in the html template still uses `publishedDate` (as any template in SearXNG does).

I agree that `published_date` would be better, but that should be done in a different PR and also update the documentation.

Bnyro commented on searx/engines/npm.py:
```diff
@@ -26,15 +24,14 @@
         }
     )
-    params['url'] = search_api + args
+    params["url"] = search_api + args
```
nit: unrelated, please remove

Bnyro commented on searx/engines/npm.py:
```diff
@@ -43,23 +40,26 @@ def response(resp):
                 "version": package.get("version"),
                 "maintainer": package.get("author", {}).get("name"),
-                'publishedDate': publishedDate,
+                "published_date": published_date,
```
```suggestion
                "publishedDate": published_date,
```

Bnyro commented on searx/engines/pypi.py:
```diff
@@ -23,44 +21,44 @@
                 "content": content,
                 "version": version,
-                'publishedDate': parser.parse(created_at),  # type: ignore
+                "published_date": parser.parse(created_at),  # type: ignore
```
```suggestion
                "publishedDate": parser.parse(created_at),  # type: ignore
```

Bnyro commented on searx/templates/simple/result_templates/packages.html:
```diff
@@ -47,17 +47,24 @@
-          <a href="{{ result.source_code_url }}" target="_blank">{{ _('Source code') }}</a>
-        {%- endif -%}
+      {% for name, link in result.links.items() %}
+        {% if not loop.first %} | {% endif %}
```
That's a nice trick, didn't know that's possible :)

Bnyro reviewed (changes requested):
I personally think it would be better to keep `source_code_url`, `documentation_url` and `homepage` normal fields (not part of the dict), because they do behave differently than normal `links`. It's a good idea to display them together, but for the developer experience, they should stay separated because one would otherwise always need to look up what specific edge cases for the `links` dictionary exist (as it's not properly documented, and doesn't follow the way other templates work).

The [GitHub engine](https://github.com/searxng/searxng/blob/58320222e2725c5e370226a1df78f9e951202e05/searx/engines/github.py#L63) would require modifications too (`source_code_url` and `homepage`) looking at the current changes. But when following the suggestion above, it won't need to be modified.

Bnyro commented on docs/dev/engines/engine_overview.rst:
```diff
@@ -556,10 +556,6 @@ type.
-     - the location of the project's source code
+   * - links
+     - :py:class:`dict`
+     - a dictionary of link name to link url, e.g. {"homepage": {http://example.com"}
```
```suggestion
     - a dictionary of link name to link url, e.g. {"homepage": "http://example.com"}
```
nit: typo

dkuku commented on searx/engines/hex.py:
```diff
@@ -46,11 +44,24 @@ def response(resp):
+        if not source_url and k.lower() in ["source", "github", "gitlab"]:
+            source_url = v
+
+    for k, v in links.items():
```
first check all the keys then if none matches check the values

return42 commented on docs/dev/engines/engine_overview.rst:
```diff
@@ -563,3 +563,7 @@ type.
+
+   * - links
+     - :py:class:`dict`
+     - additional links in the form of ``{'link_name': 'http://example.com'}``
```
I fixed it to rendered in a ``monospace font``

return42 commented on searx/templates/simple/result_templates/packages.html:
```diff
@@ -58,6 +58,14 @@
+            <a href="{{ link }}" target="_blank">
+              {{- _(name) -}}
+            </a>
+          {%- endfor -%}
```
I added some `{%-` & `-%}` to get rid of not needed whitespaces in the output.

dkuku replied:
Should this name be normalised: upper or lower?

return42 replied:
Good question .. I hesitate to normalize in the templates .. upper-/lowercase depends on the l10n .. we have languages that do not follow the rules of the alphabet (chinese, hebraic, arabic ..) .. in my opinion, a template should be neutral with regard to i18n.

Normalization should therefore better be done in the engine ... if hex does not return a normalized value (if some results contain “Author” and others “author”) we would have to normalize this in the hex engine ... I pushed https://github.com/searxng/searxng/pull/3464/commits/5269dbdd053cc57c5cb40e4a606a7c2351730687 on top of this branch.

Is it OK to merge this commit with your PR here?

dkuku replied:
it's ok.

return42 commented on searx/templates/simple/result_templates/packages.html:
```diff
@@ -58,6 +58,14 @@
+          {%- for name, link in result.links.items() -%}
+            {% if not loop.first or result.homepage or result.source_code_url %} | {% endif %}
+            <a href="{{ link }}" target="_blank">
+              {{- _(name) -}}
```
argueable .. but if we have a translation, then lets translate. Common words like `author` are already in our translation memory ..

EN ("Author"):

![image](https://github.com/searxng/searxng/assets/554536/9a986f04-b52a-43b6-9cda-ffe4b54f24bc)


DE ("Autor"):  
![image](https://github.com/searxng/searxng/assets/554536/079b8db6-c317-4a97-9501-96432286f4c9)

Common terms can be added to [searx/searxng.msg](https://github.com/searxng/searxng/blob/master/searx/searxng.msg) where the translation of `author` also comes from (a term that is frequently used in social media):

https://github.com/searxng/searxng/blob/645a840d8211a016d808cf7da1ab62ff12d1db56/searx/searxng.msg#L87-L97

return42 reviewed (approved):
@dkuku @Bnyro I squashed the commit history into one commit and slightly changed the commit message.  And I did some small modifications / see [diff](https://github.com/searxng/searxng/compare/bb1d3e3b6fc54c1515909877a4b31344855bcbb6..7dc6a368b69dfe766d32542d1714e31d8dcc1a95#diff-60332926fe56944d856be975ba056cddb5ebeaf7370dc02df2bfdb17e11ad480R569)

See my comments in this review below for more details .. if you are OK with what I have done, give a short feedback to merge this PR.

Again, thanks a lot for your work on this PR :+1:

return42 committed:
[mod] hex engine: normalize (some of) the linked terms

The names of the links are rather tags than real names, and they sometimes vary
greatly in their spelling:

- GitHub: github, Github
- Source code: Repository, SCM, Project Source Code
- Documentation: docs, Documentation

It was standardized to terms such as 'Source code' and 'Documentation', as
translations already exist for these terms.

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>
